src/beast/beast/wsproto/detail/mask.h

```diff
@@ -0,0 +1,396 @@
+    }
+}
+
+template<class _>
+void
+maskgen_t<_>::rekey()
+{
+    n_ = 0;
+    std::random_device rng;
+    std::array<std::uint32_t, 32> e;
+    for(auto& i : e)
+        i = rng();
+    std::seed_seq ss(e.begin(), e.end());
+    g_.seed(ss);
+}
+
+using maskgen = maskgen_t<>;
+
+//------------------------------------------------------------------------------
+
+//using prepared_key_type = std::size_t;
+using prepared_key_type = std::uint32_t;
+//using prepared_key_type = std::uint64_t;
+
+inline
+void
+prepare_key(std::uint32_t& prepared, std::uint32_t key)
+{
+    prepared = key;
+}
+
+inline
+void
+prepare_key(std::uint64_t& prepared, std::uint32_t key)
+{
+    prepared =
+        (static_cast<std::uint64_t>(key) << 32) | key;
+}
+
+template<class T>
+inline
+std::enable_if_t<std::is_integral<T>::value, T>
+rol(T t, unsigned n = 1)
+{
+    auto constexpr bits =
+        static_cast<unsigned>(
+            sizeof(T) * CHAR_BIT);
+    n &= bits-1;
+    return static_cast<T>((t << n) |
+        (static_cast<std::make_unsigned_t<T>>(t) >> (bits - n)));
+}
+
+template <class T>
+inline
+std::enable_if_t<std::is_integral<T>::value, T>
+ror(T t, unsigned n = 1)
+{
+    auto constexpr bits =
+        static_cast<unsigned>(
+            sizeof(T) * CHAR_BIT);
+    n &= bits-1;
+    return static_cast<T>((t << (bits - n)) |
+        (static_cast<std::make_unsigned_t<T>>(t) >> n));
+}
+
+// 32-bit Uuoptimized
+//
+template<class = void>
+void
+mask_inplace_safe(
+    boost::asio::mutable_buffer const& b,
+        std::uint32_t& key)
+{
+    using boost::asio::buffer_cast;
+    using boost::asio::buffer_size;
+    auto n = buffer_size(b);
+    auto p = buffer_cast<std::uint8_t*>(b);
+    for(auto i = n / sizeof(key); i; --i)
+    {
+        *p ^=  key      ; ++p;
+        *p ^= (key >> 8); ++p;
+        *p ^= (key >>16); ++p;
+        *p ^= (key >>24); ++p;
+    }
+    n %= sizeof(key);
+    switch(n)
+    {
+    case 3: p[2] ^= (key >>16);
+    case 2: p[1] ^= (key >> 8);
+    case 1: p[0] ^=  key;
+        key = ror(key, n*8);
+    default:
+        break;
+    }
+}
+
+// 64-bit unoptimized
+//
+template<class = void>
+void
+mask_inplace_safe(
+    boost::asio::mutable_buffer const& b,
+        std::uint64_t& key)
+{
+    using boost::asio::buffer_cast;
+    using boost::asio::buffer_size;
+    auto n = buffer_size(b);
+    auto p = buffer_cast<std::uint8_t*>(b);
+    for(auto i = n / sizeof(key); i; --i)
+    {
+        *p ^=  key      ; ++p;
+        *p ^= (key >> 8); ++p;
+        *p ^= (key >>16); ++p;
+        *p ^= (key >>24); ++p;
+        *p ^= (key >>32); ++p;
+        *p ^= (key >>40); ++p;
+        *p ^= (key >>48); ++p;
+        *p ^= (key >>56); ++p;
+    }
+    n %= sizeof(key);
+    switch(n)
+    {
+    case 7: p[6] ^= (key >>16);
+    case 6: p[5] ^= (key >> 8);
+    case 5: p[4] ^=  key;
+    case 4: p[3] ^= (key >>24);
+    case 3: p[2] ^= (key >>16);
+    case 2: p[1] ^= (key >> 8);
+    case 1: p[0] ^=  key;
```

Comment:
This looks very suspicious!  The top 3 shift amounts are either wrong or need a comment explaining why they are right.